Add atomicDEX support for Evrmore Coin (EVR)

hans-schmidt opened the pull request:
Add atomicDEX support for EVR
Two ElectrumX servers each support TCP, SSL, and wss (websockets) uinsg Let'sEncrypt certificates.
See [https://evrmorecoin.org](https://evrmorecoin.org ) for info about the project.

cipig commented:
can you please move the icon to the icons_original directory? the icons dir will be filled by a script after merge

i fear the blockexplorer will need some tweaking, because it doesn't accept requests of the form https://evr.cryptoscope.io/tx/23042ce12206ec845eb70099f09b39505f11be0a0ba1abe6900cf75d91369aab
or do you have one that uses the format `/tx/<txid>` and `/address/<address>` ?

smk762 commented:
i the suffixes are different, they can be defined in https://github.com/KomodoPlatform/coins/blob/master/explorers/explorer_paths.json

hans-schmidt commented:
Thanks.
I moved the EVR icon into the icons_original directory and added an item with correct paths in explorer_paths.json

cipig commented:
i added it to Desktop and wanted to try a swap... it failed with timeout from electrums

i saw that the 2 electrums are not in sync:
```
(echo '{ "id": 1, "method": "blockchain.headers.subscribe" }'; sleep 1) | ncat electrum1-mainnet.evrmorecoin.org 50001
{"jsonrpc":"2.0","result":{"hex":"0000003012059e6d95331db9de5088980b8fa297c827405806efb92483f4220000000000e145ba44922a88f8631c770534fbf48f578b7fbeffc0256eb7acbecf5dfaefac25dc326432b0751b8d820300314f844200306f0acf8ea25730ffd5fa8bf960d14624a5f252e731f15a8ff168c0296c9a793ca437","height":230029},"id":1}

(echo '{ "id": 1, "method": "blockchain.headers.subscribe" }'; sleep 1) | ncat electrum2-mainnet.evrmorecoin.org 50001
{"jsonrpc":"2.0","result":{"hex":"00000030a0ea58cb58ea1eb2e500d3db1a5947ecf306ab5cb1de6038840a1c000000000092730559ed655e14c28a85ce5eb6db636f53e1fd264906931041d1b7db3c0b4650fd30641281001c997a03001e90140900402316e63cae4d8343d94f4106da940133921e9174254ebb6c92a9e5a9c40d70671b38","height":227993},"id":1}
```

so electrum2 is like 200 blocks behind and electrum1 is also not on same height as https://evr.cryptoscope.io/ (230154)

hans-schmidt commented:
Sorry about that. Murphy's law at work.

After adding Let's Encrypt and websocket support to the servers recently, I added a linux ulimit memory limit to both of them to make sure that they didn't crash their VMs if there was any problem. Looks like I set the ulimit too low, so the C++ blockchain nodes behind the ElectrumX servers both crashed due to out-of-memory. Prior to the ulimit, one of the servers had been running with ElectrumX since February. 

I have removed the memory limits and restarted the nodes. Tomorrow, I will re-check the servers, verify that an mm2 swap works, and set up server monitoring to alert me of any problems. If it happens again, I will move them to more powerful VMs.

cipig commented:
electrums are now working and on same height as explorer
funny thing is that my address already had some balance on EVR... i can't remember where
`{"coin":"EVR","balance":"2787.77664296","unspendable_balance":"0","address":"Eb25ERw8Sy24SRqDgPmv6QWcnQrCVmTx9i"}`
but i get an error when i try to use that balance, i get `the transaction was rejected by network rules.\\n\\n16: bad-txns-spend-of-expired-airdrop`
the explorer does not know my address
question is how i can get rid of that balance, because ADEX will try to spend it and run into the error from the chain
will try to delete my TX_CACHE, maybe it goes away

cipig commented:
that didn't help, i still see that balance and ADEX tries to use it: `{"coin":"EVR","balance":"2787.77664296","unspendable_balance":"0","address":"Eb25ERw8Sy24SRqDgPmv6QWcnQrCVmTx9i"}`

cipig commented:
activated on a second VPS and that also has balance: `{"coin":"EVR","balance":"5494.83918626","unspendable_balance":"0","address":"ESt1e1WEB3mQ2mAW6mpoFjECqa56XUNBWe"}`
where does this balance come from?
problem is that ADEX tries to spend it, but can't

hans-schmidt commented:
I will investigate today and also try to do a swap. I saw none of this strange behavior when I successfully did two EVR swaps some weeks ago. Sorry this has been such a time sink.

cipig commented:
> I will investigate today and also try to do a swap. I saw none of this strange behavior when I successfully did two EVR swaps some weeks ago. Sorry this has been such a time sink.

It will work fine with fresh addresses that don't have those airdropped coins
what are those? where do they come from? can they be removed from balance completely? they can't be spent anyway, at tx with those leads to `bad-txns-spend-of-expired-airdrop` error from the chain

hans-schmidt commented:
I suspect what is happening is the following:
EVR is a fork of RVN. For that reason we share a common HD derivation path. We eventually plan to move to a unique path, but there will probably always be EVR on the original path. So it is possible that you have a pre-existing EVR balance because those are RVN coins which were "cloned" as a result of the chain fork. This is no different than for any UTXO chain fork.

However, the way the EVR chaiin fork was implemented was somewhat unique. RVN's history included an inflation bug which created coins, and it also had a large number of "lost" coins and a huge database size, all of which EVR did not want to inherit. So at RVN block height 2,510,000 a snapshot was taken of the top 50,000 RVN addresses with the largest balances (over 99% of all RVN). This list was copied into the EVR genesis block as an airdrop and Evrmore was started as essentially a new chain. Everyone was instructed that they needed to move their airdropped EVR to a new address (or the same address as a new UTXO) before EVR chain height 86,400. At that EVR blockchain height, any unmoved EVR airdrop UTXO's "expired" (ie became unspendable and essentially burned). This removed all of the problematic coins from RVN's history, resulting in over 5 billion coins burned. Trying to spend those burned airdrop coins will give the chain node error message which you are seeing.

cipig commented:
> At that EVR blockchain height, any unmoved EVR airdrop UTXO's "expired" (ie became unspendable and essentially burned).

that explains it... my utxos are exactly those expired ones
can the wallet be changed so that it excludes those utxos? problem is that everybody who has such expired utxos from the RVN airdrop will not be able to sell EVR or send a EVR transaction because ADEX will try to use those expired utxos and get the error

cipig commented:
ADEX does not have coin control feature added... wouldn't be easy anyway, since during a swap, you don't send txes manually, they are done by ADEX in the background
question is what to do with EVR... we can merge this, but all people that have expired utxos from the airdrop in their address will get the error and not be able to do swaps... looks like all my 6 market maker nodes have those, so i won't be able to do market making

hans-schmidt commented:
I understand. I think that we will be able to fix this on our end fairly quickly. I would prefer that the merge continue so that we can get things moving, especially since the mobile app can take some time for playstore approval.

I am evaluating the fix in our C++ node wallet. But even more quickly, it appears easy to modify the ElectrumX servers to ignore those genesis block UTXOs as unspendable (which it already does on some other chains). I will be trying that out and testing it over the next few days.

cipig reviewed (approved):
![image](https://user-images.githubusercontent.com/32116761/231159723-645a0d35-ced7-4ed0-bb28-a9def018c157.png)

explorer links are working too
